src/Microsoft.AspNetCore.Blazor.Server/MonoDebugProxy/MonoDebugProxyAppBuilderExtensions.cs

```diff
@@ -0,0 +1,178 @@
+    {
+        public static void UseMonoDebugProxy(this IApplicationBuilder applicationBuilder)
+        {
+            applicationBuilder.UseWebSockets();
+
+            applicationBuilder.Use((context, next) =>
+            {
+                var requestPath = context.Request.Path;
+                if (!requestPath.StartsWithSegments("/_framework/debug"))
+                {
+                    return next();
+                }
+
+                if (requestPath.Equals("/_framework/debug/ws-proxy", StringComparison.OrdinalIgnoreCase))
+                {
+                    return DebugWebSocketProxyRequest(context);
+                }
+
+                if (requestPath.Equals("/_framework/debug", StringComparison.OrdinalIgnoreCase))
+                {
+                    return DebugHome(context);
+                }
+
+                context.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                return Task.CompletedTask;
+            });
+        }
+
+        private static async Task DebugWebSocketProxyRequest(HttpContext context)
+        {
+            if (!context.WebSockets.IsWebSocketRequest)
+            {
+                context.Response.StatusCode = 400;
+                return;
+            }
+
+            var browserUri = new Uri(context.Request.Query["browser"]);
+            await new MonoProxy().Run(context, browserUri);
+        }
+
+        private static async Task DebugHome(HttpContext context)
+        {
+            context.Response.ContentType = "text/html";
+
+            var request = context.Request;
+            var appRootUrl = $"{request.Scheme}://{request.Host}{request.PathBase}/";
+            var targetTabUrl = request.Query["url"];
+            if (string.IsNullOrEmpty(targetTabUrl))
+            {
+                context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                await context.Response.WriteAsync("No value specified for 'url'");
+                return;
+            }
+
+            // TODO: Allow overriding port (but not hostname, as we're connecting to the
+            // local browser, not to the webserver serving the app)
+            var debuggerHost = "http://localhost:9222";
+            var debuggerTabsListUrl = $"{debuggerHost}/json";
+            IEnumerable<BrowserTab> availableTabs;
+
+            try
+            {
+                availableTabs = await GetOpenedBrowserTabs(debuggerHost);
+            }
+            catch (Exception ex)
+            {
+
+                await context.Response.WriteAsync($@"
+    <h1>Unable to find debuggable browser tab</h1>
+    <p>
+        Could not get a list of browser tabs from <code>{debuggerTabsListUrl}</code>.
+        Ensure Chrome is running with debugging enabled.
+    </p>
+    <h2>Resolution</h2>
+    {GetLaunchChromeInstructions(appRootUrl)}
+    <p>... then use that new tab for debugging.</p>
+    <h2>Underlying exception:</h2>
+    <pre>{ex}</pre>
```

Comment:
It would have been nice to use server-side Razor Pages to implement this basic built-in UI, but it seemed like using precompiled Razor views in a NuGet package involves conforming to Razor Pages file layout and routing conventions (e.g., having a `Pages` directory) which is not useful here.

We can maybe change this if we end up implementing more sophisticated built-in server-side UIs.

Reply:
I think this is fine for now. We don't want to actually bring in MVC and the MVC view engine into this part of the codes.